language-plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -0,0 +1,615 @@
+-- -------------------------------------------------------------------------------------------------
+-- [infer| G !- iwrap pat arg term : ifix vPat vArg]
+inferTypeM (IWrap ann pat arg term) = do
+    k <- inferKindM arg
+    checkKindOfPatternFunctorM ann pat k
+    vPat <- normalizeTypeM $ void pat
+    vArg <- normalizeTypeM $ void arg
+    checkTypeM ann term =<< unfoldFixOf vPat vArg k
+    pure $ TyIFix () <$> vPat <*> vArg
+
+-- [infer| G !- term : ifix vPat vArg]    [infer| G !- vArg :: k]
+-- -----------------------------------------------------------------------
+-- [infer| G !- unwrap term : NORM (vPat (\(a :: k) -> ifix vPat a) vArg)]
+inferTypeM (Unwrap ann term)        = do
+    vTermTy <- inferTypeM term
+    case unNormalized vTermTy of
+        TyIFix _ vPat vArg -> do
+            k <- inferKindM $ ann <$ vArg
+            -- Subparts of a normalized type, so normalized.
+            unfoldFixOf (Normalized vPat) (Normalized vArg) k
+        _                  -> throwError (TypeMismatch ann (void term) (TyIFix () dummyType dummyType) vTermTy)
+-- [check| G !- ty :: *]    ty ~>? vTy
+-- -----------------------------------
+-- [infer| G !- error ty : vTy]
+inferTypeM (Error ann ty)           = do
+    checkKindM ann ty $ Type ()
+    normalizeTypeM $ void ty
+
+inferTypeM (Let ann1 recurs bs inTerm) = do
+    tyInTerm <- case recurs of
+        -- [check| G !- ty :: *]  [infer| G,withBindsLinear(b) !- inT : ty] (forall b in bs. [checkWellFormed| Glinear !- b]) 
+        -- -------------------------------------------------
+        -- [infer| G !- (let (nonrec) bs in inT) :: ty
+        NonRec ->
+            foldr
+               (\b acc -> do
+                   checkWellformBind NonRec b
+                   -- linear scope: the new identifiers will be visible in the next binding
+                   withBind b acc)
+               (inferTypeM inTerm)
+               bs
+        -- [check| G !- ty :: *]  G'=G,withBinds(bs)  [infer| G' !- inT : ty] (forall b in bs. [checkWellFormed| G' !- b])
+        -- -------------------------------------------------
+        -- [infer| G !- (let (rec) bs in inT) :: ty
+        Rec -> do
+            -- Check that there are no conflicting new identifiers (same-name)
+            -- that are introduced by the bindings of this the letrec
+            -- separately check the VarTypes environment for conflicts
+            -- TODO: move these 2 checks to a different pass, e.g. in renaming phase
+            assertNoRecConflict $ foldMap bindingNewIds bs
+            -- separately check the VarkInds environment for conflicts
+            assertNoRecConflict . mapMaybe bindingNewTyId $ toList bs
+            -- now with all bindings in the env, check for wellformedness
+            withBinds bs $ do
+              checkWellformBinds bs
+              inferTypeM inTerm
+    -- In paper's Fig.6, T-Let and T-LetRec rules dictate that: G !- inTerm :: *
```

Comment:
This comment reads oddly to me. It sounds like you're saying that it's a bug in the rules that we get scoping errors, when in fact that's the entire point of this rule. So it's odd to say it's unnecessary because all terms have types of kind *, since the whole point is to check the scoping.

Reply:
What I wanted to say is that the paper says `Γ |- Τ :: *` which if I want to implement it in PIR,  I have to translate it to `[checkKind| G, binds(b) | inTerm :: *` . I need some extra scope to check the kind of inTerm, whereas the paper states that i do not need any extra scope (Γ).